packages/block-editor/src/components/block-controls/index.js

```diff
@@ -11,24 +11,69 @@ import {
 	__experimentalToolbarContext as ToolbarContext,
 	createSlotFill,
 	ToolbarGroup,
+	__experimentalUseSlot as useSlot,
 } from '@wordpress/components';
 
 /**
  * Internal dependencies
  */
 import useDisplayBlockControls from '../use-display-block-controls';
 
-const { Fill, Slot } = createSlotFill( 'BlockControls' );
+const BlockControlsDefault = createSlotFill( 'BlockControls' );
+const BlockControlsBlock = createSlotFill( 'BlockControlsBlock' );
+const BlockControlsInline = createSlotFill( 'BlockFormatControls' );
+const BlockControlsOther = createSlotFill( 'BlockControlsOther' );
 
-function BlockControlsSlot( props ) {
+const groups = {
+	default: BlockControlsDefault,
+	block: BlockControlsBlock,
+	inline: BlockControlsInline,
+	other: BlockControlsOther,
+};
+
+const slotNames = {
+	default: 'BlockControls',
+	block: 'BlockControlsBlock',
+	inline: 'BlockFormatControls',
+	other: 'BlockControlsOther',
+};
+
+function BlockControlsSlot( { group = 'default', ...props } ) {
 	const accessibleToolbarState = useContext( ToolbarContext );
-	return <Slot { ...props } fillProps={ accessibleToolbarState } />;
+	const Slot = groups[ group ].Slot;
+	const slot = useSlot( slotNames[ group ] );
+	const hasFills = Boolean( slot.fills && slot.fills.length );
+
+	if ( ! hasFills ) {
+		return null;
+	}
+
+	if ( group === 'default' ) {
+		return (
+			<Slot
+				{ ...props }
+				bubblesVirtually
+				fillProps={ accessibleToolbarState }
+			/>
+		);
+	}
+
+	return (
+		<ToolbarGroup>
```

Comment:
Should we add labels to the groups? It could be a good way to improve navigation when using screen-readers.

Reply:
Good question, I don't know. When we navigate, we generally end up directly in the button. So I'm not sure if it's useful.